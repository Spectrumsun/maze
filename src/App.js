import React, { useState } from 'react';
import axios from 'axios'
import './App.css';

export const fetchData = async (width, height) => {
  try {
    const api = `https://hubbado-maze-api.herokuapp.com/generate?width=${width}&height=${height}`;
    const fetchData = await axios.get(api)
    return fetchData.data.cells;
  }catch(error) {
    return error;
  }
}

export const calculateBody = (cell) => {
  const cellWall = [8, 4, 2, 1];
  const removeHighestCellWall = cellWall.filter((value) => value <= cell);
  let total = 0;
  const newCell = [];
  for (let i = 0; i < removeHighestCellWall.length; i++) {
    if (total === cell) {
      break;
    } else if (total < cell) {
      total += removeHighestCellWall[i];
      newCell.push(removeHighestCellWall[i]);
    } else {
      newCell.pop();
      total -= removeHighestCellWall[i];
      newCell.push(removeHighestCellWall[i]);
    }
  }
  return newCell;
};

export const convertArrayToWall = (cell) => {
  const cellBody = {
    8: 'Left',
    4: 'Bottom',
    2: 'Right',
    1: 'Top',
  };
  const cel = cell.map((value) => cellBody[value]);
  return cel;
};


export const formatCellGrid = (cells) => {
  return cells.map((value) => {
    return value.map((c) => {
      return convertArrayToWall(calculateBody(c));
    });
  });
}


const App = () => {
  const [inputValue, setInput] = useState({
    width: '',
    height: '',
  })
  const [apiData, setApiData] = useState([])
  const [status, setButton] = useState(false)

  const handleInput = (e) => {
    e.preventDefault();
    setButton(false);
    const  { name, value } = e.target;
    setInput({ ...inputValue, [name]: value });
  };

  const handleClick = async (e) => {
    const { width, height } = inputValue;
    e.preventDefault();
    setButton(true);
    const fetch = await fetchData(width, height);
    const data = formatCellGrid(fetch);
    setApiData(data);
    const gridContainer = document.querySelector('.grid-container');
    gridContainer.style.gridTemplateColumns = `repeat(${inputValue.width}, 1fr)`;
  }

  const renderMaze = () => {
     return apiData.map((cell) => {
        return cell.map((cel, index) => {
          const borderStyle = cel.reduce((obj, item) => {
            return {
                  ...obj,
                  [`border${item}Color`]: 'black',
                };
              }, {})
          return (
          <div className="maze-box" key={index} style={borderStyle}></div>
          )
        }
      )
    })
  }
  

  
  return (
      <div className="container">
        <h1>Maze</h1>
        <form className="input">
        <label>Columns</label>
        <input 
          type='number'
          name="height"
          onChange={handleInput}
          min="1"
          required
          value={inputValue.height}
        />
        <label>Rows</label>
        <input 
          type='number'
          name="width"
          onChange={handleInput}
          min="1"
          required
          value={inputValue.width}
        />
      <button
      onClick={handleClick}
      disabled={status}
      >Go!</button>
        </form>
    <div className="grid-container">
      {renderMaze()}
    </div>
    </div>
  );
}


export default App;
