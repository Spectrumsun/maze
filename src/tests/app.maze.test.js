import React from 'react';
import {shallow } from 'enzyme';
import axios from 'axios'
import App, { 
    fetchData, 
    calculateBody, 
    convertArrayToWall, 
    formatCellGrid
} from '../App';

const wrapper = shallow(<App />);
jest.mock('axios');
const data = {
  data: {
    cells: [
        [11, 11, 9, 7, 11],
        [10, 8, 4, 7, 10],
        [12, 2, 13, 3, 10],
        [13, 4, 1, 2, 10],
        [13, 5, 6, 12, 6],
    ]
  }
};

describe('Function test <Unit test>', () => {
  it('should test calculateBody function', () => {
    const cB = calculateBody(10)
     expect(cB).toEqual([8, 2]);
   });
 
   it('should test calculateBody function', () => {
     const cB = calculateBody(1)
      expect(cB).toEqual([1]);
    });
 
    it('should test convertArrayToWall function', () => {
     const cAt = convertArrayToWall([8,2,1]);
      expect(cAt).toEqual(['Left', 'Right', 'Top']);
    });
 
    it('should test convertArrayToWall function', () => {
     const cAt = convertArrayToWall([2,1]);
      expect(cAt).toEqual(['Right', 'Top']);
    });
 
    it('should test formatCellGrid function', () => {
     const cAt = formatCellGrid([[1,2,3], [2,4,8]]);
      expect(cAt).toEqual([
       [ [ 'Top' ], [ 'Right' ], [ 'Right', 'Top' ] ],
       [ [ 'Right' ], [ 'Bottom' ], [ 'Left' ] ]
     ] );
    });
 
    it('should test fetchData function', async () => {
     axios.get.mockImplementationOnce(() => Promise.resolve(data));
     await expect(fetchData(5, 5)).resolves.toEqual(data.data.cells);
     expect(axios.get).toHaveBeenCalledWith(
      `https://hubbado-maze-api.herokuapp.com/generate?width=${5}&height=${5}`
    );
    });
});

describe('<LApp />', () => {
  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should have an input field', () => {
    expect(wrapper.find('input[name="height"]').length).toEqual(1);
  });

  it('should have an input field', () => {
    expect(wrapper.find('input[name="width"]').length).toEqual(1);
  });

  it('should have an submit button', () => {
    expect(wrapper.find('button').length).toEqual(1);
  });

  it('should set the input value on change event with', () => {
    wrapper.find('input[name="height"]').simulate('change', {
      preventDefault: jest.fn(),
      target: { name: 'height', value: 5 }
    });
    expect(wrapper.find('input[name="height"]').prop('value')).toEqual(
      5,
    );
  });


  it('should set the input value on change event with', () => {
    wrapper.find('input[name="height"]').simulate('change', {
      preventDefault: jest.fn(),
      target: { name: 'width', value: 6 }
    });
    expect(wrapper.find('input[name="width"]').prop('value')).toEqual(
      6,
    );
  });


  it('should click the submit button', async () => {
    jest.spyOn(document, 'querySelector').mockReturnValueOnce({ 
      style: {
        gridTemplateColumns : 1, 
    }
    });
    axios.get.mockImplementationOnce(() => data);
    wrapper.find('button').simulate('click', {
      preventDefault: jest.fn(),
    });

    axios.get.mockImplementationOnce(() => data);
    expect(wrapper.find('button').prop('disabled')).toEqual(true);
  });


  it('should click render maze',  () => {
    jest.spyOn(document, 'querySelector').mockReturnValueOnce({ 
      style: {
        gridTemplateColumns : 1, 
    }
    });
    axios.get.mockImplementationOnce(() => data);
    wrapper.find('button').simulate('click', {
      preventDefault: jest.fn(),
    });
    axios.get.mockImplementationOnce(() => data);
    wrapper.update()
    expect(wrapper.find('.maze-box').length).toEqual(25);
  });
});
