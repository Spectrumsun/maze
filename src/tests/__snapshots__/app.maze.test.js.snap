// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LApp /> should match the snapshot 1`] = `"<div class=\\"container\\"><h1>Maze</h1><form class=\\"input\\"><label>Columns</label><input type=\\"number\\" name=\\"height\\" min=\\"1\\" required=\\"\\" value=\\"\\"/><label>Rows</label><input type=\\"number\\" name=\\"width\\" min=\\"1\\" required=\\"\\" value=\\"\\"/><button>Go!</button></form><div class=\\"grid-container\\"></div></div>"`;
